### PR TITLE
impl: add HTTP status code handling

### PR DESCRIPTION
**Pull request by scotthart**
This PR adds only the HTTP status code handling.

This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/7870)

### CODE REVIEW

File: google/cloud/internal/curl_handle_factory.cc [hunk 1]
```diff
@@ -159,6 +159,10 @@ void PooledCurlHandleFactory::SetCurlOptions(CURL* handle) {
   }
 }
 
+CurlHandle GetCurlHandle(std::shared_ptr<CurlHandleFactory> const& factory) {
```
Comment (coryan) [inline google/cloud/internal/curl_handle_factory.cc:162]: Seems line this could be a tiny PR by itself too?

**Review by coryan — COMMENTED** [review]
1,000 lines is a bit much for me to grok.  Consider first a PR with just the status code mapping, which is more controversial than I thought.

File: google/cloud/internal/rest_response.cc [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+  if (HttpStatusCode::kMinContinue <= code &&
+      code < HttpStatusCode::kMinSuccess) {
+    return StatusCode::kOk;
+  }
+  // We treat the 200s as Okay results.
+  if (HttpStatusCode::kMinSuccess <= code &&
+      code < HttpStatusCode::kMinRedirects) {
+    return StatusCode::kOk;
+  }
+  if (code == HttpStatusCode::kResumeIncomplete) {
```
[the remark this answers is not in the document]
Comment (scotthart) [reply google/cloud/internal/rest_response.cc:42]: Added code to populate the original http status code in the returned Status.

[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/internal/rest_response.cc:42]: But what are we going to do about the `308` -> `kFailedPrecondition`?  Is that the right mapping for other services?

File: google/cloud/internal/rest_response.cc [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+namespace google {
+namespace cloud {
+namespace rest_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace {
+// Maps HTTP status codes to enumerators in `StatusCode`. The code here is
+// organized by increasing range (or value) of the errors, just to make it
+// readable. There are probably shorter (and/or more efficient) ways to write
+// this, but we went for readability.
+StatusCode MapHttpCodeToStatus(long code) {  // NOLINT(google-runtime-int)
```
Comment (coryan) [inline google/cloud/internal/rest_response.cc:30]: Most of the code in this function is not tested.  That is Okay if you are planning to add the tests as part of the next PR

File: google/cloud/internal/rest_response.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/internal/rest_response.cc:42]: Maybe just add a bug/issue to track this and let's merge this PR while we figure it out?

**Review by coryan — APPROVED** [review]

**Pull request merged**